README.md
Federated single sign-out for SAMLP providers from express.js applications.

![Build Status](https://github.com/auth0/samlp-logout/workflows/Tests/badge.svg)

## Installation

```
npm i samlp-logout --save
```

## Usage

```javascript
var SamlpLogout = require('samlp-logout');

var logout = SamlpLogout({
  issuer: 'urn:foobar',
  identityProviderUrl: 'http://xx.b',
  identityProviderSigningCert: fs.readFileSync('idpSigningKey.pem'), // validate LogoutRequest signature
  key: fs.readFileSync('signingKey.key'), // sign LogoutResponse
  cert: fs.readFileSync('signingKey.pem')
});

// assuming we have req.samlSessionIndex and req.samlNameID
app.get('/logout', function (req, res, next) {
  req.samlSessionIndex = 'abc';
  req.samlNameID = {
    value: 'xyz',
    Format: 'urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress'
  };
  next();
}, logout);
```

## License

MIT - 2014 - AUTH0 INC.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [4.0.0](https://github.com/auth0/samlp-logout/compare/v3.1.0...v4.0.0) (2023-12-19)

- Properly calling next() callback when using HTTP-POST bindings

## [3.1.0](https://github.com/auth0/samlp-logout/compare/v3.0.0...v3.1.0) (2021-08-30)

- Do not send empty RelayState fields. This works around a bug in Azure AD.

## [3.0.0](https://github.com/auth0/samlp-logout/compare/v2.3.3...v3.0.0) (2021-02-09)

### ⚠ BREAKING CHANGES

- since 4.0.0, next() callback is properly called with HTTP-POST bindings

- newer mochas drop support for older node versions and only support 10 upwards

- fix npm audit warnings for cheerio, mocha & xml-crypto ([02bed18](https://github.com/auth0/samlp-logout/commit/02bed1893d44879e8a1ed306dd1d460df4d6554f))


test/unit.tests.js
const expect      = require('chai').expect;
const url         = require('url');
const DOMParser   = require('xmldom').DOMParser;
const fs          = require('fs');
const path        = require('path');
const zlib        = require('zlib');
const samlpLogout = require('..');

const credentials = {
  cert: fs.readFileSync(path.join(__dirname, 'fixture', 'samlp.test-cert.pem')),
  key:  fs.readFileSync(path.join(__dirname, 'fixture', 'samlp.test-cert.key')),
};

describe('samlpLogout', function () {
  describe('with a malformed key', function () {
    it('should invoke the callback with an error', function (done) {
      const logout = samlpLogout({
        issuer: 'http://example.org',
        identityProviderUrl: 'http://myadfs.com/ls',
        cert: credentials.cert,
        key:  'badly-formatted-key',
        relayState: 'foo=bar',
        deflate: true,
        identityProviderSigningCert: credentials.cert // only for testing purposes
      });

      logout({
        samlNameID: '112233',
        samlSessionIndex: '554433',
        query: {},
        body: {}
      }, {
        redirect: function (location) {
          const parsedUrl = url.parse(location, true);

          expect(parsedUrl.protocol).to.equal('http:');
          expect(parsedUrl.hostname).to.equal('myadfs.com');
          expect(parsedUrl.pathname).to.equal('/ls');

          zlib.inflateRaw(new Buffer(parsedUrl.query.SAMLRequest, 'base64'), function (err, buffer) {
            if (err) return done(err);

            SAMLRequestWithDeflate = parsedUrl.query.SAMLRequest;
            SAMLRequest = new DOMParser().parseFromString(buffer.toString());
            RelayState = parsedUrl.query.RelayState;
            SigAlg = parsedUrl.query.SigAlg;
            Signature = parsedUrl.query.Signature;
          });
        }
      }, (err) => {
        expect(err.message).to.equal('error:0909006C:PEM routines:get_name:no start line')
        done();
      });
    });
  });
});


test/utils.tests.js
const expect = require('chai').expect;

const utils = require('../lib/utils');

describe('utils', function () {
    describe('generateUniqueID', function() {
        it('should generate an ID 20 chars long', function() {
            expect(utils.generateUniqueID().length).to.equal(20);
        });
    });
    describe('generateUniqueID', function() {
        it('should generate an ID from the alphabet', function() {
            expect('abcdef0123456789'.split('')).to.include.members(utils.generateUniqueID().split(''));
        });
    });
});

test/signature.tests.js
var expect = require('chai').expect;
var url = require('url');
var samlpLogout = require('../');
var xmldom = require('xmldom');
var DOMParser = xmldom.DOMParser;

var fs = require('fs');
var path = require('path');

var select = require('xml-crypto').xpath;
var SignedXml = require('xml-crypto').SignedXml;
var FileKeyInfo = require('xml-crypto').FileKeyInfo;
var fs = require('fs');


var credentials = {
  cert: fs.readFileSync(path.join(__dirname, 'fixture', 'samlp.test-cert.pem')),
  key:  fs.readFileSync(path.join(__dirname, 'fixture', 'samlp.test-cert.key')),
};

describe('samlp logout signature', function () {
  var parsedUrl, xml, SAMLRequest;

  before(function (done) {
    var logout = samlpLogout({
      issuer: 'http://example.org',
      identityProviderUrl: 'http://myadfs.com',
      cert: credentials.cert,
      key:  credentials.key
    });

    logout({
      samlNameID: '112233',
      samlSessionIndex: '554433'
    }, {
      redirect: function (location) {
        parsedUrl = url.parse(location, true);
        xml = new Buffer(parsedUrl.query.SAMLRequest, 'base64').toString();
        SAMLRequest = new DOMParser().parseFromString(xml);
        done();
      }
    });
  });

  it('should be valid', function () {
    var signature = select(SAMLRequest, "/*/*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']")[0];
    var sig = new SignedXml();

    sig.keyInfoProvider = new FileKeyInfo(path.join(__dirname, 'fixture', 'samlp.test-cert.pem'));
    sig.loadSignature(signature.toString());
    sig.checkSignature(xml);

    expect(sig.validationErrors).to.be.empty;
  });
});

lib/utils.js
const url = require('url');
const xtend = require('xtend');

const IdGenerator = require('auth0-id-generator');
const generator = new IdGenerator({len: 20, alphabet: "abcdef0123456789"});

/**
 * Generates a random alphanumeric ID.
 * > IDs are 20 characters long and use the hexadecimal alphabet
 * @returns {string}
 */
exports.generateUniqueID = function () {
  return generator.get();
};

exports.getRoundTripDateFormat = function() {
  //http://msdn.microsoft.com/en-us/library/az4se3k1.aspx#Roundtrip
  const date = new Date();
  return date.getUTCFullYear() + '-' +
        ('0' + (date.getUTCMonth()+1)).slice(-2) + '-' +
        ('0' + date.getUTCDate()).slice(-2) + 'T' +
        ('0' + date.getUTCHours()).slice(-2) + ":" +
        ('0' + date.getUTCMinutes()).slice(-2) + ":" +
        ('0' + date.getUTCSeconds()).slice(-2) + "Z";
};

exports.appendQueryString = function(initialUrl, query) {
  const parsed = url.parse(initialUrl, true);
  parsed.query = xtend(parsed.query, query);
  delete parsed.search;
  return url.format(parsed);
};


lib/signers.js
var SignedXml = require('xml-crypto').SignedXml;
var xmlCrypto = require('xml-crypto');
var crypto    = require('crypto');
var encoders  = require('./encoders');

var algorithms = {
  signature: {
    'rsa-sha256': 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256',
    'rsa-sha1':  'http://www.w3.org/2000/09/xmldsig#rsa-sha1'
  },
  digest: {
    'sha256': 'http://www.w3.org/2001/04/xmlenc#sha256',
    'sha1': 'http://www.w3.org/2000/09/xmldsig#sha1'
  }
};

var DEFAULT_SIG_ALG = 'rsa-sha256';
var DEFAULT_DIGEST_ALG = 'sha256';

module.exports.getSigAlg = function (options) {
  return algorithms.signature[options.signatureAlgorithm || DEFAULT_SIG_ALG];
};

module.exports.signXml = function (options, xml) {
  var signatureAlgorithm = options.signatureAlgorithm || DEFAULT_SIG_ALG;
  var digestAlgorithm = options.digestAlgorithm || DEFAULT_DIGEST_ALG;

  var sig = new SignedXml(null, {
    signatureAlgorithm: algorithms.signature[signatureAlgorithm],
    idAttribute: 'ID'
  });

  sig.addReference(options.reference || "//*[local-name(.)='LogoutRequest' and namespace-uri(.)='urn:oasis:names:tc:SAML:2.0:protocol']",
                  ["http://www.w3.org/2000/09/xmldsig#enveloped-signature",
                   "http://www.w3.org/2001/10/xml-exc-c14n#"],
                  algorithms.digest[digestAlgorithm]);

  sig.signingKey = options.key;

  var pem = encoders.removeHeaders(options.cert);
  sig.keyInfoProvider = {
    getKeyInfo: function () {
      return "<X509Data><X509Certificate>" + pem + "</X509Certificate></X509Data>";
    }
  };

  sig.computeSignature(xml, {
    location: {
      reference: "//*[local-name(.)='Issuer']",
      action: 'after'
    }
  });

  return sig.getSignedXml();
};

module.exports.validateXmlEmbeddedSignature = function (xml, options) {
  var signature = xmlCrypto.xpath(xml, "/*/*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']")[0];
  var sig = new xmlCrypto.SignedXml();
  sig.keyInfoProvider = {
    getKeyInfo: function () {
      return '<X509Data></X509Data>';
    },
    getKey: function () {
      return options.identityProviderSigningCert;
    }
  };

  sig.loadSignature(signature.toString());
  sig.checkSignature(xml.toString());

  return sig.validationErrors;
};

module.exports.sign = function (options, content) {
  var signatureAlgorithm = options.signatureAlgorithm || DEFAULT_SIG_ALG;
  var signer = crypto.createSign(signatureAlgorithm.toUpperCase());
  signer.update(content);
  return signer.sign(options.key, 'base64');
};

module.exports.isValidContentAndSignature = function (content, signature, options) {
  var verifier = crypto.createVerify(options.signatureAlgorithm.split('#')[1].toUpperCase());
  verifier.update(content);
  return verifier.verify(options.identityProviderSigningCert, signature, 'base64');
};


lib/sign_xml.js
var SignedXml = require('xml-crypto').SignedXml;
var encoders = require('./encoders');

var algorithms = {
  signature: {
    'rsa-sha256': 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256',
    'rsa-sha1':  'http://www.w3.org/2000/09/xmldsig#rsa-sha1'
  },
  digest: {
    'sha256': 'http://www.w3.org/2001/04/xmlenc#sha256',
    'sha1': 'http://www.w3.org/2000/09/xmldsig#sha1'
  }
};

module.exports = function (options, xml) {
  var signatureAlgorithm = options.signatureAlgorithm || 'rsa-sha256';
  var digestAlgorithm = options.digestAlgorithm || 'sha256';

  var sig = new SignedXml(null, {
    signatureAlgorithm: algorithms.signature[signatureAlgorithm],
    idAttribute: 'ID'
  });

  sig.addReference("//*[local-name(.)='LogoutRequest' and namespace-uri(.)='urn:oasis:names:tc:SAML:2.0:protocol']",
                  ["http://www.w3.org/2000/09/xmldsig#enveloped-signature",
                   "http://www.w3.org/2001/10/xml-exc-c14n#"],
                  algorithms.digest[digestAlgorithm]);

  sig.signingKey = options.key;

  var pem = encoders.removeHeaders(options.cert);
  sig.keyInfoProvider = {
    getKeyInfo: function () {
      return "<X509Data><X509Certificate>" + pem + "</X509Certificate></X509Data>";
    }
  };

  sig.computeSignature(xml, "//*[local-name(.)='Issuer']");
  return sig.getSignedXml();
};

index.js
var zlib      = require('zlib');
var DOMParser = require('xmldom').DOMParser;
var xpath     = require('xpath');
var util      = require('util');
var qs        = require('querystring');
var url       = require('url');
var templates = require('./templates');
var trim_xml  = require('./lib/trim_xml');
var signers   = require('./lib/signers');
var utils     = require('./lib/utils');

var BINDINGS = {
  HTTP_POST:      'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST',
  HTTP_REDIRECT:  'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
};

var RESPONSE_EMBEDDED_SIGNATURE_PATH = "//*[local-name(.)='LogoutResponse']/*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']";
var REQUEST_EMBEDDED_SIGNATURE_PATH = "//*[local-name(.)='LogoutRequest']/*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']";

function prepareAndSendToken (req, res, type, token, options, cb) {
  var send = function (params) {
    if (options.protocolBinding === BINDINGS.HTTP_POST) {
      // HTTP-POST
      res.set('Content-Type', 'text/html');
      
      var payload = {
        type:         type,
        callback:     options.identityProviderUrl,        
        token:        params[type]
      };
      if('RelayState' in params) {
        payload.RelayState = params.RelayState;
      }
      
      return res.send(templates.Form(payload));
    }

    // HTTP-Redirect
    var samlResponseUrl = utils.appendQueryString(options.identityProviderUrl, params);
    res.redirect(samlResponseUrl);
  };

  var params = {};
  params[type] = null;
  params.RelayState = req.body.RelayState || req.query.RelayState || options.relayState || '';
  if (params.RelayState === '') {
    // if there is no RelayState value, the parameter should be omitted from the signature computation
    delete params.RelayState;
  }

  // canonical request
  token = trim_xml(token);

  if (options.protocolBinding === BINDINGS.HTTP_POST || !options.deflate) {
    // HTTP-POST or HTTP-Redirect without deflate encoding
    try {
      token = signers.signXml(options, token);
    } catch (err) {
      return cb(err);
    }

    params[type] = new Buffer(token).toString('base64');
    const sendResult = send(params);
    cb();
    return sendResult;
  }

  // Default: HTTP-Redirect with deflate encoding (http://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf - section 3.4.4.1)
  zlib.deflateRaw(new Buffer(token), function (err, buffer) {
    if (err) return cb(err);

    params[type] = buffer.toString('base64');

    // construct the Signature: a string consisting of the concatenation of the SAMLResponse,
    // RelayState (if present) and SigAlg query string parameters (each one URLencoded)    
    params.SigAlg = signers.getSigAlg(options);

    try {
      params.Signature = signers.sign(options, qs.stringify(params));
    } catch (e) {
      return cb(e);
    }

    send(params);
    cb();
  });
}

function isTokenExpired (logoutNode) {
  var notOnOrAfterText = logoutNode.getAttribute('NotOnOrAfter');
  if (notOnOrAfterText) {
    var notOnOrAfter = new Date(notOnOrAfterText);
    notOnOrAfter = notOnOrAfter.setMinutes(notOnOrAfter.getMinutes() + 10); // 10 minutes clock skew
    var now = new Date();
    return now > notOnOrAfter;
  }

  return false;
}

function validateSignature (req, type, xml, options) {
  var isRequestSigned = req.body[type] ?
    xpath.select(REQUEST_EMBEDDED_SIGNATURE_PATH, xml).length > 0 : !!req.query.SigAlg;

  if (isRequestSigned) {
    if (req.body[type] || !options.deflate) {
      // HTTP-POST or HTTP-Redirect without deflate encoding
      var validationErrors = signers.validateXmlEmbeddedSignature(xml, options);
      if (validationErrors && validationErrors.length > 0) {
        throw new Error(validationErrors.join('; '));
      }
    }
    else {
      // HTTP-Redirect with deflate encoding
      // Section 3.4.4.1 https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf

      // Get unescaped query string values, signature must be verified using the original
      // escaping rules.
      var parsedUrl = url.parse(req.url);
      var rawQuery = qs.parse(
        parsedUrl.query,
        null,
        null,
        // Do not decode query values
        { decodeURIComponent: function(s) { return s; } }
      );

      if (!rawQuery.SigAlg) {
        throw new Error('SigAlg parameter is mandatory');
      }

      var payload = (
        type + '=' + rawQuery[type] +
        (rawQuery.RelayState ? '&RelayState=' + rawQuery.RelayState : '') +
        '&SigAlg=' + rawQuery.SigAlg
      );

      var valid = signers.isValidContentAndSignature(payload, req.query.Signature, {
        identityProviderSigningCert: options.identityProviderSigningCert,
        signatureAlgorithm: req.query.SigAlg
      });

      if (!valid) {
        throw new Error('invalid signature: the signature value ' + req.query.Signature + ' is incorrect');
      }
    }
  } else if (type === 'SAMLRequest') {
    throw new Error('LogoutRequest message MUST be signed when using an asynchronous binding (POST or Redirect)');
  }
}

module.exports = function (options) {
  options = options || {};

  // Scenario #2: IdP Single Logout - SLO
  var idpSingleLogOut = function (req, res, next) {
    var SAMLRequest = req.query.SAMLRequest || req.body.SAMLRequest;

    var validateAndRespond = function (err, buffer) {
      if (err) return next(err);

      var xml = new DOMParser().parseFromString(buffer.toString());
      var logoutRequestNode = xpath.select("//*[local-name(.)='LogoutRequest']", xml)[0];

      // validate expiration
      if (isTokenExpired(logoutRequestNode)) {
        return next(new Error('LogoutRequest has expired'));
      }

      // validate signature
      try {
        validateSignature(req, 'SAMLRequest', xml, options);
      } catch (e) {
        return next(e);
      }

      // get ID, Issuer, NameID and SessionIndex
      var parsedRequest = {};
      parsedRequest.id = logoutRequestNode && logoutRequestNode.getAttribute('ID');

      var issuerNode = xpath.select("//*[local-name(.)='Issuer']", xml);
      parsedRequest.issuer = issuerNode && issuerNode[0] && issuerNode[0].textContent;

      var nameIdNode = xpath.select("//*[local-name(.)='NameID']", xml);
      parsedRequest.nameId = nameIdNode && nameIdNode[0] && nameIdNode[0].textContent;

      var sessionIndexNode = xpath.select("//*[local-name(.)='SessionIndex']", xml);
      parsedRequest.sessionIndex = sessionIndexNode && sessionIndexNode[0] && sessionIndexNode[0].textContent;

      // validate parameters (NameID and SessionIndex)
      if (!parsedRequest.sessionIndex) { return next(new Error('Missing SessionIndex')); }
      if (!parsedRequest.nameId) { return next(new Error('Missing NameID')); }

      var checkSessionIndex = typeof options.validSessionIndex === 'function';
      var isValidSessionIndex = checkSessionIndex && options.validSessionIndex(parsedRequest);
      if (checkSessionIndex && !isValidSessionIndex) {
        return next(new Error('Invalid SessionIndex/NameID'));
      }

      // prepare and send response
      var logoutResponse = templates.LogoutResponse({
        ID: utils.generateUniqueID(),
        IssueInstant: utils.getRoundTripDateFormat(),
        Destination: options.identityProviderUrl,
        InResponseTo: parsedRequest.id,
        Issuer: options.issuer,
        StatusCode: 'urn:oasis:names:tc:SAML:2.0:status:Success' // TODO: support all status codes
      });

      options.reference = "//*[local-name(.)='LogoutResponse' and namespace-uri(.)='urn:oasis:names:tc:SAML:2.0:protocol']";

      prepareAndSendToken(req, res, 'SAMLResponse', logoutResponse, options, next);
    };

    if (req.body.SAMLRequest || !options.deflate) {
      // HTTP-POST or HTTP-Redirect without deflate encoding
      return validateAndRespond(null, new Buffer(SAMLRequest, 'base64'));
    }

    // Default: HTTP-Redirect with deflate encoding
    zlib.inflateRaw(new Buffer(SAMLRequest, 'base64'), validateAndRespond);
  };

  // Scenario #1.2: parse and validate SAMLResponse (Logout Response)
  var parseResponse = function (req, res, next) {
    var SAMLResponse = req.query.SAMLResponse || req.body.SAMLResponse;

    var parseAndValidate = function (err, buffer) {
      if (err) return next(err);

      var xml = new DOMParser().parseFromString(buffer.toString());
      var parsedResponse = {};

      // status code
      var statusCodes = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:protocol', 'StatusCode');
      var statusCodeXml = statusCodes[0];
      if (statusCodeXml) {
        parsedResponse.status = statusCodeXml.getAttribute('Value');

        // status sub code
        var statusSubCodeXml = statusCodes[1];
        if (statusSubCodeXml) {
          parsedResponse.subCode = statusSubCodeXml.getAttribute('Value');
        }
      }

      // status message
      var samlStatusMsgXml = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:protocol', 'StatusMessage')[0];
      if (samlStatusMsgXml) {
        parsedResponse.message = samlStatusMsgXml.textContent;
      }

      // status detail
      var samlStatusDetailXml = xml.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:protocol', 'StatusDetail')[0];
      if (samlStatusDetailXml) {
        parsedResponse.detail = samlStatusDetailXml.textContent;
      }

      req.parsedSAMLResponse = parsedResponse;

      // validate signature
      try {
        validateSignature(req, 'SAMLResponse', xml, options);
      } catch (e) {
        return next(e);
      }

      // validate status
      if (parsedResponse.status !== 'urn:oasis:names:tc:SAML:2.0:status:Success') {
        var err_message = parsedResponse.message && parsedResponse.detail ?
          util.format('%s (%s)', parsedResponse.message, parsedResponse.detail) :
          parsedResponse.message ||
          parsedResponse.detail ||
          util.format('unexpected SAMLP Logout response (%s)', parsedResponse.status);

        return next(new Error(err_message));
      }

      next();
    };

    if (req.body.SAMLResponse || !options.deflate) {
      // HTTP-POST or HTTP-Redirect without deflate encoding
      return parseAndValidate(null, new Buffer(SAMLResponse, 'base64'));
    }

    // Default: HTTP-Redirect with deflate encoding
    zlib.inflateRaw(new Buffer(SAMLResponse, 'base64'), parseAndValidate);
  };

  return function (req, res, next) {
    req.body = req.body || {};
    req.query = req.query || {};

    // validations
    if (!options.cert || !options.key) {
      return next(new Error('signing key is mandatory (options.cert and options.key)'));
    }

    if (!options.identityProviderSigningCert) {
      return next(new Error('options.identityProviderSigningCert is mandatory'));
    }

    if (req.query.SAMLRequest || req.body.SAMLRequest) {
      // Scenario #2 (IdP-Initiated SLO): parse and validate SAMLRequest and return a SAMLResponse
      return idpSingleLogOut(req, res, next);
    }

    if (req.query.SAMLResponse || req.body.SAMLResponse) {
      // Scenario #1.2: parse and validate the Logout Response
      return parseResponse(req, res, next);
    }

    // Scenario #1.1: initialize a Logout Request
    var logoutRequest = templates.LogoutRequest({
      ID: utils.generateUniqueID(),
      IssueInstant: utils.getRoundTripDateFormat(),
      Issuer: options.issuer,
      NameID: typeof req.samlNameID === 'string' ? { value: req.samlNameID } : req.samlNameID,
      SessionIndex: req.samlSessionIndex,
      Destination: options.identityProviderUrl
    });

    prepareAndSendToken(req, res, 'SAMLRequest', logoutRequest, options, next);
  };
};
